#50877: add support to postgresql_privs to use "FOR { ROLE | USER } target_role" in "ALTER DEFAULT PRIVILEGES"

tcraxs opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
* add support to postgresql_privs to use "FOR { ROLE | USER } target_role" in "ALTER DEFAULT PRIVILEGES"
* Fixes #50877 
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
ATM only a workaround (See [#32987)](https://github.com/ansible/ansible/pull/32987#issuecomment-451026860) exists to use the [FOR { ROLE | USER } target_role [, ...]](https://www.postgresql.org/docs/current/sql-alterdefaultprivileges.html).
I implement a new parameter **target_roles** like the roles parameter to configure the target_role in the default privilege context.
If you don't set the target_roles or either not use the [FOR ROLE ...] option, the default privileges will be set for the user who is running the command (this will be the login_user or if you use become_user in your playbook this user).

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
- postgresql_privs

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

A possible playbook could look like this:
```paste below
- name: set default privs
  become_user: postgres
  postgresql_privs:
    db: library
    objs: TABLES,SEQUENCES
    privs: SELECT
    type: default_privs
    role: group_reader
    role_target: reader
    schema: reader
```

tcraxs committed:
#50877:
* add support to postgresql_privs to use "FOR { ROLE | USER } target_role"
   in "ALTER DEFAULT PRIVILEGES"

tcraxs committed:
fix sanity errors

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -70,6 +70,10 @@
     required: yes
+  target_roles:
+    description:
+      - Comma separated list of role (user/group) names to set as the default permissions.
```
Based on official documentation from [postgresql.org],(https://www.postgresql.org/docs/9.4/sql-alterdefaultprivileges.html,)
I suggest to change the description:
```
Comma-separated list of existing role (user/group) names to set as the default permissions
for database objects subsequently created by them.
```
Or something similar

tcraxs replied:
Thanks for the hint, I changed that and add examples to the code, how to use target_roles

Andersson007 commented:
1. "role_target" is in ADDITIONAL INFORMATION, "target_roles" is in DOCUMENTATION block of source code.
Please, change it and add several items with target_roles to EXEMPLE block of source doce (for both revoke and grand cases)

2. When I executed this playbook, test1 can see content of tables created by test_role
```
- name: test
  postgresql_privs:
    db: test
    state: absent
    objs: TABLES,SEQUENCES
    privs: SELECT
    type: default_privs
    role: test1
    target_roles: test_role
    schema: public
```
Please, check all use cases for right work.
3. Ensure that target_roles does make sense with any existing parameter.

tcraxs committed:
#50877: fix documentation and add a check for correct usage
of target_roles

tcraxs commented:
> 1. "role_target" is in ADDITIONAL INFORMATION, "target_roles" is in DOCUMENTATION block of source code.

I'm not sure what do you mean with 
> "role_target" is in ADDITIONAL INFORMATION

>    Please, change it and add several items with target_roles to EXEMPLE block of source doce (for both revoke and grand cases)

Added new EXEMPLE block for grant and revoke cases.

> Please, check all use cases for right work.
> 3. Ensure that target_roles does make sense with any existing parameter.

I added a check, that allows `target_roles` only with `type: default_privs`. As far as I see, `target_roles` will only used with ÀLTER DEFAULT PRIVILEGES`

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -70,6 +70,11 @@
+    description:
+      - Comma-separated list of existing role (user/group) names to set as the
+        default permissions for database objects subsequently created by them.
+    version_added: 2.8
```
Please, related you last changes, also change the parameter description "Comma-separated list of existing role (user/group) names to set as the default permissions for database objects subsequently created by them. I(target_roles) is only available with I(type=default_privs)."

tcraxs replied:
@Andersson007 thanks for your good hints. I forgot the "state=absent" part for this case. I added it and also added your suggestion for the description.

dagwieers replied:
Please add parameter types to every parameter.
Also, `version_added` are strings.
And it's better to use multiple items in the description if parts are unrelated.
In this case this would be better as:
```yaml
  target_roles:
    description:
    - A list of existing role (user/group) names to set as the
      default permissions for database objects subsequently created by them.
    - Parameter I(target_roles) is only available with C(type=default_privs).
    type: list
    version_added: '2.8'
```

Andersson007 commented:
I tried it:
```
 14   - name: set default privs
 15     postgresql_privs:
 16       db: mydb
 17       state: present
 18       objs: TABLES,SEQUENCES
 19       privs: SELECT
 20       type: default_privs
 21       role: test_priv2
 22       target_roles: test_role
 23       schema: public
```
it works, but then I revoke privs by playbook below:
```
 14   - name: set default privs
 15     postgresql_privs:
 16       db: mydb
 17       state: absent
 18       objs: TABLES,SEQUENCES
 19       privs: SELECT
 20       type: default_privs
 21       role: test_priv2
 22       target_roles: test_role
 23       schema: public
```
It returns that nothing has changed and test_prive can continue to read the content of new tables created by test_role after that. Shortly, seems state=absent doesn't work.
If I do "ALTER DEFAULT PRIVILEGES FOR ROLE test_role IN SCHEMA public REVOKE SELECT ON TABLES FROM test_priv2;" manually, it works as expected.

Please, check it. Also you should fix shippable tests errors.

tcraxs committed:
#50877: fix missing absent option for default privs with target_role

tcraxs committed:
#50877: add clear description, when target_roles can be used

tcraxs commented:
ready_for_review

tcraxs commented:
!needs_rebase

Andersson007 reviewed (approved):
I tried it manually. It works as expected. Good job, thank you.
I hope other contributors will approve it soon

@kostiantyn-nemchenko @andytom @astorath

shipit

Andersson007 commented:
@tcraxs if you want to help to improve Postgres support in the next release, https://github.com/ansible/community/wiki/PostgreSQL there are links to several new modules that are waiting for you ;) You can see it in "New PostgreSQL modules" section.
Also we need ci tests for existing modules. You can see them also there.

Would you mind if I add you in this working group ?

Thank you.

andytom commented:
lgtm
shipit

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -265,6 +271,35 @@
+# GRANT SELECT privileges for new TABLES objects created by librarian as
+# default to the role reader.
+# For specific
+- postgresql_privs:
```
Examples should have a `name:` entry.
That is what we recommend to users, so we should set a good example in the docs.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -265,6 +271,35 @@
+
 """
 
 import traceback
```
Traceback should not be needed, Ansible adds traceback information on errors in except-blocks automatically.

Andersson007 replied:
@tcraxs these formatting fixes are all that remain. then merge, I hope. There are some issues accumulated for postgresql_privs that I want to start to solve after the merge of this PR. So, please, don't postpone it ;)

Andersson007 replied:
I can start now but, I'm afraid, there will be many merge conflicts with your PR that you / or I need to solve.

tcraxs replied:
Sorry I was not able to get some free time for this. I will resolve the merge conflicts right now.
Regarding the "import traceback", I think that is out of my scope. I'm not sure how to resolve that, because it was already in the code before my change. I think we have to rewrite a lot of the code to remove the traceback-thing and port it to match it with the Ansible default error-handling.

Andersson007 commented:
@tcraxs please, fix conflicts, formatting, and add a changelog fragment https://docs.ansible.com/ansible/devel/community/development_process.html#making-your-pr-merge-worthy

tcraxs committed:
#50877: fix conflicts, formatting, and add a changelog fragment

tcraxs committed:
#50877: fix sanity error E335

tcraxs commented:
ready_for_review

tcraxs commented:
bot_status

Andersson007 commented:
@dagwieers , about the examples without 'name', types, traceback, etc. - I'm going to fix it in a separate PR later. If you don't have any notes, please, merge it.

Andersson007 commented:
@tcraxs thank you for this PR. Hope it will be merged.

We also have new several modules and PR that's waiting for review, you can find links to them here https://github.com/ansible/community/wiki/PostgreSQL.
You can write yourself in the table on the working group page. Welcome!

Andersson007 commented:
@chamini2 , could you please look at this PR and approve it by shipit or make notes for this

chamini2 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -679,9 +727,15 @@ def build(self):
-            self.query.append(
-                'ALTER DEFAULT PRIVILEGES IN SCHEMA {0} REVOKE ALL ON {1} FROM {2};'.format(self._schema, obj,
-                                                                                            self._for_whom))
+            if not self._as_who:
```
I would swap the conditionals: `if self._as_who: ... else: `.

Thinking in the current `else` branch is kind of weird: `not not self._as_who`

chamini2 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -617,12 +653,19 @@ def manipulate_privs(self, obj_type, privs, objs, roles,
+        if target_roles:
+            as_who = ','.join(pg_quote_identifier(r, 'role') for r in target_roles)
+        else:
+            as_who = target_roles
```
```suggestion
          as_who = None
          if target_roles:
            as_who = ','.join(pg_quote_identifier(r, 'role') for r in target_roles)
```

This `as_who = target_roles` is confusing, since `target_roles` is None, empty string or something falsy in general.

chamini2 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -698,13 +752,28 @@ def add_grant_option(self):
-                'ALTER DEFAULT PRIVILEGES IN SCHEMA {0} GRANT {1} ON {2} TO {3}'.format(self._schema, self._set_what,
-                                                                                        obj,
-                                                                                        self._for_whom))
+            if not self._as_who:
```
Same as above: swap conditionals.

chamini2 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -698,13 +752,28 @@ def add_grant_option(self):
             self.add_grant_option()
-        self.query.append(
-            'ALTER DEFAULT PRIVILEGES IN SCHEMA {0} GRANT USAGE ON TYPES TO {1}'.format(self._schema, self._for_whom))
+        if not self._as_who:
```
Same as above: swap conditionals.

chamini2 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -719,9 +788,15 @@ def build_absent(self):
-                self.query.append(
-                    'ALTER DEFAULT PRIVILEGES IN SCHEMA {0} REVOKE ALL ON {1} FROM {2};'.format(self._schema, obj,
-                                                                                                self._for_whom))
+                if not self._as_who:
```
Same as above: swap conditionals.

chamini2 commented on lib/ansible/modules/database/postgresql/postgresql_privs.py:
```diff
@@ -860,11 +936,24 @@ def main():
                     module.warn("Role '%s' does not exist, nothing to do" % roles[0].strip())
 
+        # check if target_roles is set with type: default_privs
+        if p.target_roles and not p.type == 'default_privs':
```
I am not sure this check is necessary, maybe add a warning of `target_roles` being ignored instead of a failure?

Andersson007 commented:
@chamini2 and we forgot about CI tests.
@tcraxs , this PR implements support of a new functionality, so you need to add CI tests to a new section in test/integration/targets/postgresql/tasks/postgresql_privs.yml by analogy.
If you have any questions about it or need some help, please, let me know.

tcraxs commented:
> @chamini2 and we forgot about CI tests.
> @tcraxs , this PR implements support of a new functionality, so you need to add CI tests to a new section in test/integration/targets/postgresql/tasks/postgresql_privs.yml by analogy.
> If you have any questions about it or need some help, please, let me know.

Thanks for the good points on swapping conditionals.
I will try my best to straggle through the CI tests, maybe someone will or must help me, because I'm not familiar with such tests.

Andersson007 commented:
@tcraxs , these are usual playbooks, see the files in test/integration/targets/postgresql/tasks/
Also you'll find the useful information in our pinboard https://github.com/ansible/community/issues/435
particularly there are several links in _dagwieers commented on Feb 1_
Try to figure it out.
if there will be difficulties, please, let me know directly by aaklychkov at mail dot ru

tcraxs committed:
#50877: swap conditions and fix error to warning msg

chamini2 commented:
shipit

chamini2 commented:
Oh, @Andersson007 is right, please add tests!

kostiantyn-nemchenko commented:
I wonder if it would be possible to add support of managing global default privileges and apply privileges for the whole database, not schema?
If `schema` is specified, the default privileges are altered for objects later created in that schema, otherwise permissions are applied for all newly created objects regardless of schema by omitting `IN SCHEMA` clause in alter statement. 
IMHO would be useful to have such opportunity, though not sure if this feature should be addressed by this PR.

tcraxs commented:
> Oh, @Andersson007 is right, please add tests!

I'm on it.

tcraxs committed:
#50877: add tests for default privileges

tcraxs committed:
#50877: fix tests for default privileges

Andersson007 commented:
sure, freebsd uses pgsql user (or something like this) instead of postgres

Andersson007 commented:
```
"msg": "relation \"pg_default_acl\" does not exist\nLINE 2:                    FROM pg_default_acl a\n                                ^\n"
```
CentOS 6
there is postgresql 9.0 - very old. You could check a version and use something instead of pg_default_acl if the version is 9.0

Andersson007 commented:
If it is not related with your patch, so just skip it by
```
# Test default_privs with target_role
- include: test_target_role.yml
postgres_version_resp.stdout is version('9.1', '>=')
```
I'm not sure about ('9.1', '>=') or ('9.1', '<=')

tcraxs committed:
#50877: fix tests for default privileges on centos 6

Andersson007 commented:
shipit

Andersson007 commented:
@chamini2 , if you're saticfied with the changes, please, write shipit to a new comment

Andersson007 commented:
Ok, @chamini2 , I saw, thank you

Andersson007 commented:
@tcraxs , thank you for this PR! hope you'll stay an active community member. You can help us by review of 51073, e.g. You also can find some interesting ideas here https://github.com/ansible/community/wiki/PostgreSQL or implement your own. I'll be glad to help you with review.
Also we have a community pinboard to communicate about postgres related topics https://github.com/ansible/community/issues/435

gundalow commented:
Merged into `devel` for release in Ansible 2.8

@tcraxs Thank you for this PR, your first or many I hope. Welcome to the Ansible Community!
@Andersson007 @chamini2 @dagwieers Thank you for the reviews.

@Andersson007 has kindly offered to raise a follow up PR to:
* Add `type:`
* Add `- name:` to all examples

See https://github.com/ansible/ansible/pull/54027 for an example of what this would look like
